sdk/clientcore/core/src/main/java/io/clientcore/core/http/models/SdkRequestContext.java

```diff
@@ -0,0 +1,160 @@
+// Licensed under the MIT License.
+
+package io.clientcore.core.http.models;
+
+import io.clientcore.core.annotations.Metadata;
+import io.clientcore.core.annotations.MetadataProperties;
+import io.clientcore.core.instrumentation.InstrumentationContext;
+import io.clientcore.core.instrumentation.logging.ClientLogger;
+import io.clientcore.core.utils.ProgressReporter;
+
+import java.util.function.Consumer;
+
+/**
+ * This class contains the request metadata that flows along with the {@link HttpRequest} as it goes through HTTP pipeline.
+ * It includes {@link RequestOptions} provided by the user along with metadata provided by client libraries and core components.
+ * <p>
+ * This class is intended to be used by client library code and HTTP pipeline components. Application developers should
+ * configure request metadata using {@link RequestOptions}.
+ * <p>
+ * This class does not allow any modifications to the request metadata provided by the user.
+ */
+@Metadata(properties = { MetadataProperties.FLUENT, MetadataProperties.IMMUTABLE })
+public final class SdkRequestContext {
+    // TODO (limolkova) do we need to make it public and probably move to another class?
+    private static final String PROGRESS_REPORTER_KEY = "progressReporter";
+    static final SdkRequestContext NONE = new SdkRequestContext(null, null, null);
+    private static final ClientLogger LOGGER = new ClientLogger(SdkRequestContext.class);
+    private final ProgressReporter childProgressReporter;
+    private final RequestOptions requestOptions;
+
+    private SdkRequestContext(RequestOptions options, ProgressReporter progressReporter,
+        InstrumentationContext instrumentationContext) {
+        this.requestOptions = new RequestOptions(options, this);
+        if (instrumentationContext != null) {
+            this.requestOptions.setInstrumentationContext(instrumentationContext);
+        }
+        this.childProgressReporter = progressReporter;
+    }
+
+    /**
+     * Creates a new instance of {@link SdkRequestContext} with the given options.
+     * @param options The {@link RequestOptions} to be used.
+     */
+    SdkRequestContext(RequestOptions options) {
+        this(options, null, null);
+    }
+
+    /**
+     * Creates a new instance of {@link SdkRequestContext} with the given options.
+     * @param options The {@link RequestOptions} to be used.
+     * @return The {@link SdkRequestContext} from the given options.
+     */
+    public static SdkRequestContext from(RequestOptions options) {
+        if (options != null && options.getRequestContext() != null) {
+            return options.getRequestContext();
+        }
+
+        return new SdkRequestContext(options);
+    }
+
+    /**
+     * Get the {@link RequestOptions} corresponding to this request context.
+     * @return The {@link RequestOptions} corresponding to this request context.
+     */
+    public RequestOptions toRequestOptions() {
+        return requestOptions;
+    }
```

Comment:
These methods to help switching from and to RequestOptions are nice! It's good to see we don't need to extend from RequestOptions anymore.